sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+        do_something_else(idx, temp);
+      });
+    });
+
+    // memory is available to be used by another allocation
+    q.submit_with_event([&](handler &cgh) {
+      cgh.depends_on(e5);
+      async_free(q, temp);
+    });
+  }
+
+  // memory pool releases memory back to the OS
+  q.wait();
+}
+----
+
+Another example of memory pool usage is described in the example below. In this
+example rather than creating an explicit memory pool the default memory pool is
+being used instead. There is also additional queue synchronization between the
+commands enqueued which would ordinarily lead to memory being released back to
+the OS, however, the threshold for the memory pool is extended to the SYCL RT
+will maintain this memory allocation, and therefore still provide the benefit of
+re-allocating memory from the memory pool.
+
+[source,c++]
+----
+int main(int argc, char *argv[])
+{
+  queue q(property::queue::in_order);
+
+  ext::oneapi::experimental::memory_pool memPool
+    = q.get_context().ext_oneapi_get_default_memory_pool(usm::alloc::device);
+
+  memPool.set_new_threshold(1024);
+  
+  {
+    // memory pool allocates memory from the OS
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for first kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // memory pool does not release memory back to the OS as it is still within
+  // the specified threshold
+  q.wait();
+
+  {
+    // memory pool re-uses previously allocated memory
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for second kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something_else(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // again memory pool does not release memory back to the OS
+  q.wait();
+}
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Memory pool
+
+This extension introduces the `memory_pool` class, which provides a handle to a
+memory pool owned by the SYCL runtime or a specific backend, and adheres to the
+SYCL common reference semantics.
```

Comment:
Should it also support APIs to get / read the properties that are used to construct the pool?

Reply:
Yeah I think we should have get/set APIs for the properties, for SYCL 2020 properties, there are common property member functions which we add to any class which can be constructed with properties, but since we're using the properties extension I'm not actually sure what the convention is for doing this, is there a good example of this I could follow?

Reply:
@aelovikov-intel, @Pennycook, and @rolandschulz are reviewing the new properties extension right now.  Can one of you propose the right convention for a class like `memory_pool` to follow?  This class takes several properties, but they are all runtime values.  What APIs should `memory_pool` expose in order to query them?  I assume it is some form of `has_property` and `get_property`?

Reply:
I think we have a choice here.

We could either expose `has_property()` and `get_property()`:

```c++
auto pool = memory_pool(..., sycl::properties{sycl::max_size(1024)});
if constexpr (pool.has_property<sycl::max_size>()) {
  assert(pool.get_property<sycl::max_size>().value == 1024);
}
```

...or we could expose a `properties()` member that returns the property list that was passed to the constructor:

```c++
auto pool = memory_pool(..., sycl::properties{sycl::max_size(1024)});
auto p = pool.properties();
if constexpr (p.contains<sycl::max_size>()) {
  assert(p.get<sycl::max_size>().value == 1024);
}
```

...and yes, I did just suggest that we replace `get_property` and `has_property` with less verbose alternatives!

I'd prefer something like the second approach, because:

1) We don't have to specify `has_property()` and `get_property()` in every class.
2) Returning a `properties` allows the developer to do anything they can do with a `properties`, not just two queries.
3) We can adopt shorter names: `has_property()`/`get_property()` make sense when we're querying an arbitrary class, but if we know we're querying a property list, we can use something like `contains`/`get`.

(3. would be longer if you're only querying one property, but much shorter if you query multiple properties or want to call some function/trait that accepts a list of properties.)

Reply:
I'm not so sure about `has`/`get` if we decide to go with duck-typing for a single property passed instead of the properties list. These names are much more likely to conflict with some specific property's methods.

Reply:
> This class takes several properties, but they are all runtime values.

Why are they? It looks like `ready_only` and `zero_init` should be compile-time only properties (which we define differently). Are we going to provide variable shortcuts for these two or not?